src/Distribution/Server/Util/Email.hs

```diff
@@ -0,0 +1,148 @@
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE OverloadedStrings #-}
+
+module Distribution.Server.Util.Email
+  ( EmailContent(..)
+  , emailContentStr
+  , emailContentLBS
+  , emailContentDisplay
+  , emailContentIntercalate
+  , emailContentUrl
+
+  -- * Rendering email content
+  , fromEmailContent
+  , toPlainContent
+  , toHtmlContent
+  ) where
+
+import qualified Data.ByteString.Lazy as Lazy (ByteString)
+import Data.List (intersperse)
+import Data.String (IsString(..))
+import Data.Text (Text)
+import qualified Data.Text as Text
+import qualified Data.Text.Lazy as TextL
+import qualified Data.Text.Lazy.Encoding as TextL
+import Distribution.Pretty (Pretty)
+import Distribution.Text (display)
+import Network.Mail.Mime
+import Network.URI (URI, uriToString)
+
+{- $setup
+>>> :set -XOverloadedStrings
+>>> import qualified Data.Text.IO as Text
+>>> import Network.URI (parseURI)
+-}
+
+data EmailContent
+  = EmailContentText Text
+  | EmailContentLink Text URI
+  | EmailContentSoftBreak
+  | EmailContentParagraph EmailContent
+  | EmailContentList [EmailContent]
+  | EmailContentConcat EmailContent EmailContent
+  deriving (Show)
+
+instance IsString EmailContent where
+  fromString = EmailContentText . Text.pack
+
+instance Semigroup EmailContent where
+  (<>) = EmailContentConcat
+
+instance Monoid EmailContent where
+  mempty = EmailContentText ""
+
+emailContentStr :: String -> EmailContent
+emailContentStr = EmailContentText . Text.pack
+
+emailContentLBS :: Lazy.ByteString -> EmailContent
+emailContentLBS = EmailContentText . TextL.toStrict . TextL.decodeUtf8
+
+emailContentDisplay :: Pretty a => a -> EmailContent
+emailContentDisplay = EmailContentText . Text.pack . display
+
+emailContentIntercalate :: EmailContent -> [EmailContent] -> EmailContent
+emailContentIntercalate x = mconcat . intersperse x
+
+emailContentUrl :: URI -> EmailContent
+emailContentUrl uri = EmailContentLink (uriToText uri) uri
+
+fromEmailContent :: EmailContent -> Alternatives
+fromEmailContent emailContent =
+  [ Part
+      { partType = contentType <> "; charset=utf-8"
+      , partEncoding = None
```

Comment:
I know that we also used `None` in the old version, but I am wondering whether it wouldn't be better to use `QuotedPrintableText` as mime-mail does itself: [Network.Mail.Mime.htmlPart](https://hackage.haskell.org/package/mime-mail-0.5.1/docs/src/Network.Mail.Mime.html#htmlPart). I am not sure we can actually generate any non-ASCII characters, so maybe it is irrelevant. But QuotedPrintableText does seem safer.

Reply:
Oh!! I should just use `htmlPart` and `plainPart` 😄